src/app/public/src/modules/table-of-contents/table-of-contents.component.ts

```diff
@@ -13,9 +22,76 @@ export class StacheTableOfContentsComponent implements StacheNav, AfterViewInit
   @Input()
   public routes: StacheNavLink[] = [];
 
-  constructor(private cdr: ChangeDetectorRef) {}
+  private headerHeight: any;
+  private documentElement: any;
+  private documentBottom: number;
+  private pageOffset: number;
+  private window: Window;
+  private activeRoute: StacheNavLink;
+
+  constructor(
+    private cdr: ChangeDetectorRef,
+    private windowRef: StacheWindowRef
+  ) {
+    this.window = this.windowRef.nativeWindow;
+    this.documentElement = this.window.document.documentElement;
+  }
 
   public ngAfterViewInit() {
     this.cdr.markForCheck();
+    this.getContentOffset();
+    this.updateActiveRoute();
+  }
+
+  @HostListener('window:scroll')
+  public onScroll() {
+    this.trackPageOffset();
+    this.updateActiveRoute();
+    this.updateView();
+  }
+
+  private trackPageOffset() {
+    // Represents top of page
+    // headerHeight buffer removes discrepency when hero or other 'header' elements exist
+    // 50px buffer aids in click-to-anchor and user visualization
+    this.pageOffset = ((this.window.pageYOffset - this.headerHeight) + 50);
+
+    // Tracks page bottom so final route can be highlighted if associated anchor provides limited content
+    // (Logic based on Angular implementation)
+    this.documentBottom = Math.round(this.documentElement.getBoundingClientRect().bottom);
+  }
+
+  private updateActiveRoute() {
+    // If scroll reaches the bottom of the page automatically assign last route as active route
+    // See documentBottom comment, above
+    // 5px padding helps to catch bottom of page in all window sizes/scales
+    if ((this.window.innerHeight + 5) >= this.documentBottom) {
+      this.activeRoute = this.routes[this.routes.length - 1];
+      return;
+    }
+
+    this.routes.map((route: StacheNavLink) => {
+      if (route.offsetTop <= this.pageOffset) {
+        this.activeRoute = route;
+      }
+    });
+  }
+
+  // Updates ng-class level trigger to show/hide blue border marking location on page
+  private updateView() {
+    this.routes.map((route: StacheNavLink) => {
```

Comment:
Using map is pretty expensive here, as it creates a new array for each time it's run.  Here we are running it on every trigger of scroll.  Can we just update the existing route/array?

Reply:
maybe just use `this.routes.forEach`?

Reply:
We can also make use of the parameters in the `forEach` call to check other elements in the routes.   Something like 
```
  private updateView() {
    this.routes.forEach((route: StacheNavLink, index: number) => {
      if (this.isActiveRoute(route, index)) {
        route.isCurrent = true;
        this.activeRoute = route;
      } else {
        route.isCurrent = false;
      }
    });
  }

  private isActiveRoute(route: StacheNavLink, index: number): boolean {
    if ((this.window.innerHeight + 5) >= this.documentBottom) {
      this.activeRoute = this.routes[this.routes.length - 1];
      return route === this.activeRoute;
    }

    return route.offsetTop <= this.pageOffset
      && (this.routes[index + 1] === undefined || this.routes[index + 1].offsetTop > this.pageOffset);
  }
```
Would allow us to eliminate the updateActiveRoute method, and make it one `forEach` loop instead of two.